numba/np/arrayobj.py

```diff
@@ -6376,3 +6376,27 @@ def take_along_axis_impl(arr, indices, axis):
             return _take_along_axis_impl(arr, indices, axis, Ni, Nk,
                                          indices_broadcast_shape)
     return take_along_axis_impl
+
+
+@overload(np.nan_to_num)
+def nan_to_num_impl(X, copy=True, nan=0.0):
+    if isinstance(X, types.Number):
+        def impl(X, copy=True, nan=0.0):
+            if np.isnan(X):
+                return nan
+            return X
+
+    elif type_can_asarray(X):
+        def impl(X, copy=True, nan=0.0):
+            X_ = np.asarray(X)
+            output = np.copy(X_) if copy else X_
+            output_flat = output.flat
+            for i in range(output.size):
+                if np.isnan(output_flat[i]):
+                    output_flat[i] = nan
+            return output
+    else:
+        raise errors.TypingError("The first argument must be a scalar or an "
+                                 "array-like")
```

Comment:
Can you add a test for this branch?

Reply:
This PR was updated to include more input validation and tests to include each code path.